[Merged by Bors] - refactor(Tactic/Util/AtLocation): allow throwing a warning when no progress is being made

grunweg opened the pull request:
A future PR will use this to make `by_contra!` and friends warn if the `push_neg` step makes no progress.
The `push_neg` and `ring_nf` tactic configuration options are already modified accordingly;
no behaviour changes are performed as part of this PR.

Co-authored by: Jovan Gerbscheid <jovan.gerbscheid@gmail.com>

---

- [x] depends on: #37898
- [x] depends on: #37999
- [x] depends on: #38008

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

When merging, all the commits will be squashed into a single commit
listing all co-authors.

Co-authors in the squash commit are gathered from two sources:

First, all authors of commits to this PR branch are included. Thus,
one way to add co-authors is to include at least one commit authored by
each co-author among the commits in the pull request. If necessary, you
may create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

Second, co-authors can also be listed in lines at the very bottom of
the commit message (that is, directly before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines
at the bottom of the commit message (before the `---`, and also before
any "Co-authored-by" lines) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

JovanGerb reviewed (approved):
Thanks for making this PR!

This PR doesn't change the current behaviour, so it would be easy to review. Though I won't maintainer merge it since I'm a co-author.

JovanGerb commented:
I have a feeling we might want to be able to customize the warning message more. We could for example have it say "use `contrapose` instead of `contrapose!`". Or something like "Use `set_option blah false` to turn off this warning".

So maybe the `.warning` constructor of your inductive type should carry some extra information?

fpvandoorn commented on Mathlib/Tactic/Ring/RingNF.lean:
```diff
@@ -151,7 +151,7 @@ elab (name := ringNF) "ring_nf" tk:"!"? cfg:optConfig loc:(location)? : tactic =
   let s ← IO.mkRef {}
   let m := AtomM.recurse s cfg.toConfig evalExpr (cleanup cfg)
-  transformAtLocation (m ·) "ring_nf" loc cfg.failIfUnchanged false
+  transformAtLocation (m ·) "ring_nf" loc (if cfg.failIfUnchanged then .error else .silent) false
```
improve configuration here as well (or add a todo?)

fpvandoorn reviewed (approved):
LGTM, mostly

thorimur commented on Mathlib/Util/AtLocation.lean:
```diff
@@ -42,16 +42,30 @@ def Lean.Elab.Tactic.withNondepPropLocation (loc : Location) (atLocal : FVarId 
 
+/-- Different settings of communicating about a tactic which made no progress: do nothing,
+print a warning or throw an error. -/
+inductive LogIfUnchanged where
```
Note that `logInfo`, `logWarning`, `logError` and such pass a `Lean.MessageSeverity` to `log`, which is either `.info`, `.warning`, or `.error`!

Hypothetically we could use an `Option MessageSeverity` to capture `silent`, but it seems an important difference here is control flow: `LogIfUnchanged` *throws* an error instead of logging it.

I don't think there's too much wrong with having a small inductive here, especially because of the readability it brings to call points: when a function takes in a `LogIfUnchanged` setting, we know exactly what it's going to use that for! :)

However, maybe `BehaviorIfUnchanged` or something to that effect could be a good tweak to the name, since we log per se only in one case.

grunweg replied:
I like `BehaviorIfUnchanged` (and went with that).

thorimur replied:
You mentioned going with `BehaviorIfUnchanged`, but it's still `LogIfUnchanged`—were some changes not pushed, maybe? (I don't mind which name we go with, just want to be sure it's what you decided! :) )

thorimur replied:
Nevermind, sorry, I missed the comment about #37899! 🤦‍♀️

grunweg commented on Mathlib/Util/AtLocation.lean:
```diff
@@ -78,35 +92,39 @@ def transformAtLocalDecl (m : Expr → ReaderT Simp.Context MetaM Simp.Result) (
-    throwError "{proc} made no progress at {ldecl.userName}"
+  if tgt.cleanupAnnotations == r.expr.cleanupAnnotations then
+    match ifUnchanged with
+    | .warning => logWarning m!"{proc} made no progress at {ldecl.userName}"
```
~~How do I do this?~~ ok, have some pointers now
Can this be a follow-up PR?

grunweg replied:
Looking into this: addTermInfo' (or addConstantInfo) seem to require the syntax where to put the hover. This is not easily available here. Am I overlooking something?

(Yes, I could probably thread it through - but that's a separate refactor now.)

thorimur replied:
This should be enough! :)
```suggestion
    | .warning => logWarning m!"`{proc}` made no progress at {ldecl.toExpr}"
```
(Note: `logWarning` adds the message context to make the free variable expression interpretable, so we shouldn't have to worry about that.)

grunweg replied:
Thanks; trying this in #37899 now.

grunweg replied:
And it seems to work :tada:

grunweg commented:
I made fixed the merge conflicts and made the requested changes in #37899. Once CI passes there, I'll mirror the changes here.

thorimur commented on Mathlib/Util/AtLocation.lean:
```diff
@@ -42,16 +42,30 @@ def Lean.Elab.Tactic.withNondepPropLocation (loc : Location) (atLocal : FVarId 
+  if unchanged then
+    match ifUnchanged with
+    | .warning => logWarning m!"{proc} made no progress on the goal"
+    | .error => throwError "{proc} made no progress on the goal"
```
Perhaps? Or too verbose?

We should also code fence now :)
```suggestion
    | .warning => logWarning m!"`{proc}` made no progress on the goal:\n{.ofGoal goal}"
    | .error => throwError "`{proc}` made no progress on the goal:\n{.ofGoal goal}"
```

grunweg replied:
#37999 adds code fences :-)

grunweg replied:
About including the goal state, I'm on the fence (no pun intended). We don't do so for other tactics (like ring or abel), so perhaps not?

thorimur replied:
I'm fine either way here. :) In principle I'm not sure how much we want to value consistency with other behavior here, since perhaps those should be changed too!

I am worried it would be too verbose, but might be informative in logs. (Another option is "`on the goal of type{indentD tgt}`".)

But it is technically out of scope for this PR, so if there's doubt or discussion needed I say let's leave it unchanged. :)

adomani replied:
I am talking about things I don't really know, but would adding the goal information be useful for LLMs?  I think that they struggle to get information from the infoview and placing it in errors is easily accessible.  🤷

grunweg replied:
Not an expert either --- but isn't the fix to that "use a proper MCP framework"?

In general, I'm fine with adding extra information to help LLM, but not when this makes live worse for humans. *Personally*, the output is too chatty for me. If you disagree, then I'm fine with adding this.

adomani replied:
I definitely agree that we should aim for human benefit and only add LLM stuff if it does not hinder the human experience, so I'm happy to go with the shorter output.

As a side remark, it might be a good idea to have a verbosity setting for diagnostics that allowed the user to control how much detail is given.  There we could have an LLM verbosity.

thorimur commented on Mathlib/Util/AtLocation.lean:
```diff
@@ -78,35 +92,39 @@ def transformAtLocalDecl (m : Expr → ReaderT Simp.Context MetaM Simp.Result) (
+  if tgt.cleanupAnnotations == r.expr.cleanupAnnotations then
+    match ifUnchanged with
+    | .warning => logWarning m!"{proc} made no progress at {ldecl.userName}"
+    | .error => throwError "{proc} made no progress at {ldecl.userName}"
```
Likewise:
```suggestion
    | .error => throwError "`{proc}` made no progress at {ldecl.toExpr}"
```

grunweg commented:
Next preliminary clean-up is at #37999.

thorimur commented:
Just to be clear, is #37899 replacing this PR? (If so, would it make sense to convert the non-preferred PR to a draft?)

grunweg committed:
refactor(AtLocation): allow throwing a warning when no progress is being made

A future PR will use this to make `by_contra!` and friends warn if the
`push_neg` step makes no progress. The `push_neg` and `ring_nf` tactic
configuration options are already modified to allow this;
no behaviour changes are performed as part of this PR.

Co-authored by: Jovan Gerbscheid <jovan.gerbscheid@gmail.com>

grunweg commented:
No: I just squashed the commits on the other PR, and decided it was easiest to overwrite this PR with it. So this PR is the current state again. (Sorry for losing history! I hope it was not too bad; the new diff is certainly straightforward to review.)

thorimur commented on Mathlib/Tactic/Ring/RingNF.lean:
```diff
@@ -38,8 +38,8 @@ inductive RingMode where
 structure Config extends AtomM.Recurse.Config where
-  /-- if true, then fail if no progress is made -/
-  failIfUnchanged := true
+  /-- how to behave if no progress is made: warn, error or keep silent -/
```
```suggestion
  /-- How to behave if no progress is made: warn, error or keep silent -/
```

thorimur commented on Mathlib/Util/AtLocation.lean:
```diff
@@ -47,18 +47,34 @@ def Lean.Elab.Tactic.withNondepPropLocation (loc : Location) (atLocal : FVarId 
+  /-- Log a warning -/
+  | warning
+  /-- Throw an error -/
+  | error
```
Some more confirmation for hoverers at the point of hovering:
```suggestion
  /-- Stay silent if this action has no effect. -/
  | silent
  /-- Log a warning if this action has no effect. -/
  | warning
  /-- Throw an error if this action has no effect. -/
  | error
```

thorimur commented on Mathlib/Util/AtLocation.lean:
```diff
@@ -47,18 +47,34 @@ def Lean.Elab.Tactic.withNondepPropLocation (loc : Location) (atLocal : FVarId 
 namespace Mathlib.Tactic
 open Lean Meta Elab.Tactic
 
+
```
```suggestion
```

thorimur reviewed (commented):
Looks good to me! Some small text-based comments to finish up. :)

maintainer delegate

grunweg committed:
Apply suggestions from code review

Co-authored-by: Thomas R. Murrills <68410468+thorimur@users.noreply.github.com>

adomani commented on Mathlib/Tactic/Ring/RingNF.lean:
```diff
@@ -38,8 +38,8 @@ inductive RingMode where
 structure Config extends AtomM.Recurse.Config where
-  /-- if true, then fail if no progress is made -/
-  failIfUnchanged := true
+  /-- How to behave if no progress is made: warn, error or keep silent -/
```
```suggestion
  /-- How to behave if no progress is made: warn, error or keep silent. Default to error -/
```
I am mostly thinking about looking at the hover and wondering: what happens if I do not provide this field?

grunweg committed:
Apply suggestion from @adomani

Co-authored-by: damiano <adomani@gmail.com>

Vierkantor reviewed (approved):
Thanks!

bors r+
